File: src/converter/plugins/convertClassDeclaration.ts
import ts, {ClassDeclaration, Declaration, ModifierLike} from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {ifPresent} from "../render.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {InheritanceModifierService, inheritanceModifierServiceKey} from "./InheritanceModifierPlugin.js";
import {DeclarationMergingService, declarationMergingServiceKey} from "./DeclarationMergingPlugin.js";
import {ConverterContext} from "../context.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {NamespaceInfoService, namespaceInfoServiceKey} from "./NamespaceInfoPlugin.js";
import {InjectionService, injectionServiceKey} from "./InjectionPlugin.js";
import {InjectionType} from "../injection.js";

function extractModifiers(member: Declaration): ModifierLike[] {
    if (
        ts.isPropertyDeclaration(member)
        || ts.isMethodDeclaration(member)
        || ts.isConstructorDeclaration(member)
        || ts.isGetAccessorDeclaration(member)
        || ts.isSetAccessorDeclaration(member)
        || ts.isIndexSignatureDeclaration(member)
    ) {
        return Array.from(member.modifiers ?? [])
    }

    return []
}

function resolveConstructors(node: ClassDeclaration, members: Declaration[], context: ConverterContext): Declaration[] {
    const constructors = members.filter(it => ts.isConstructorDeclaration(it))
    if (constructors.length > 0) return constructors

    if (!node.heritageClauses) return constructors

    const heritageClause = node.heritageClauses.find(it => it.token === ts.SyntaxKind.ExtendsKeyword)
    if (!heritageClause) return constructors

    const parentReference = heritageClause.types[0].expression

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    const typeChecker = typeScriptService?.program.getTypeChecker()
    const parentSymbol = typeChecker?.getSymbolAtLocation(parentReference)
    if (!parentSymbol) return constructors

    const parentDeclaration = parentSymbol.valueDeclaration
    if (!parentDeclaration || !ts.isClassDeclaration(parentDeclaration)) return constructors

    const declarationMergingService = context.lookupService<DeclarationMergingService>(declarationMergingServiceKey)
    const namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    const resolveNamespaceStrategy = namespaceInfoService?.resolveNamespaceStrategy?.bind(namespaceInfoService)

    const mergedMembers = declarationMergingService?.getMembers(parentDeclaration, resolveNamespaceStrategy) ?? Array.from(parentDeclaration.members)

    return resolveConstructors(parentDeclaration, mergedMembers, context)
}

export const convertClassDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isClassDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    const declarationMergingService = context.lookupService<DeclarationMergingService>(declarationMergingServiceKey)
    declarationMergingService?.cover(node)

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    const namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)
    const inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
    const injectionService = context.lookupService<InjectionService>(injectionServiceKey)

    const exportModifier = node.modifiers?.find(it => it.kind === ts.SyntaxKind.ExportKeyword)
    exportModifier && checkCoverageService?.cover(exportModifier)

    const declareModifier = node.modifiers?.find(it => it.kind === ts.SyntaxKind.DeclareKeyword)
    declareModifier && checkCoverageService?.cover(declareModifier)

    const name = (node.name && render(node.name)) ?? "Anonymous"

    const inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)
    const injections = injectionService?.resolveInjections(node, InjectionType.MEMBER, context, render)
    const staticInjections = injectionService?.resolveInjections(node, InjectionType.STATIC_MEMBER, context, render)
    const heritageInjections = injectionService?.resolveInjections(node, InjectionType.HERITAGE_CLAUSE, context, render)

    const namespace = typeScriptService?.findClosest(node, ts.isModuleDeclaration)

    let externalModifier = "external "

    if (namespace !== undefined && namespaceInfoService?.resolveNamespaceStrategy(namespace) === "object") {
        externalModifier = ""
    }

    const typeParameters = (declarationMergingService?.getTypeParameters(node) ?? node.typeParameters)
        ?.map(typeParameter => render(typeParameter))
        ?.filter(Boolean)
        ?.join(", ")

    const heritageClauses = (declarationMergingService?.getHeritageClauses(node) ?? node.heritageClauses)
        ?.map(heritageClause => render(heritageClause))
        ?.filter(Boolean)
        ?.join(", ")

    const injectedHeritageClauses = heritageInjections
        ?.filter(Boolean)
        ?.join(", ")

    const fullHeritageClauses = [heritageClauses, injectedHeritageClauses]
        .filter(Boolean)
        .join(", ")

    const resolveNamespaceStrategy = namespaceInfoService?.resolveNamespaceStrategy?.bind(namespaceInfoService)

    const mergedMembers = declarationMergingService?.getMembers(node, resolveNamespaceStrategy) ?? node.members

    // cover private members
    mergedMembers
        .filter(member => extractModifiers(member).some(it => it.kind === ts.SyntaxKind.PrivateKeyword))
        .forEach(member => checkCoverageService?.cover(member))

    const publicMembers = mergedMembers
        .filter(member => extractModifiers(member).every(it => it.kind !== ts.SyntaxKind.PrivateKeyword))

    const constructors = resolveConstructors(node, publicMembers, context)
    const otherMembers = publicMembers.filter(it => !ts.isConstructorDeclaration(it))

    const members = [...constructors, ...otherMembers]
        .filter(member => extractModifiers(member).every(it => it.kind !== ts.SyntaxKind.StaticKeyword))
        .map(member => render(member))
        .join("\n")

    const staticMembers = otherMembers
        .filter(member => extractModifiers(member).some(it => it.kind === ts.SyntaxKind.StaticKeyword))
        .map(member => render(member))
        .join("\n")

    const injectedMembers = (injections ?? [])
        .join("\n")

    const staticInjectedMembers = (staticInjections ?? [])
        .join("\n")

    let companionObject = ""

    if (staticMembers.length > 0) {
        companionObject = "\n" + `
        
companion object {
${staticMembers}${ifPresent(staticInjectedMembers, it => `\n${it}`)}
}
        `.trim()
    }

    return `
${ifPresent(inheritanceModifier, it => `${it} `)}${externalModifier}class ${name}${ifPresent(typeParameters, it => `<${it}>`)}${ifPresent(fullHeritageClauses, it => ` : ${it}`)} {
${members}${ifPresent(injectedMembers, it => `\n${it}`)}${companionObject}
}
    `.trim()
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertClassDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.InjectionType
import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import js.array.ReadonlyArray
import typescript.*

private fun extractModifiers(member: Declaration): ReadonlyArray<ModifierLike> {
    if (isPropertyDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()
    if (isMethodDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()
    if (isConstructorDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()
    if (isGetAccessorDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()
    if (isSetAccessorDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()
    if (isIndexSignatureDeclaration(member)) return member.modifiers?.asArray() ?: emptyArray()

    return emptyArray()
}

private fun resolveConstructors(
    node: ClassDeclaration,
    members: ReadonlyArray<Declaration>,
    context: Context,
): ReadonlyArray<Declaration> {
    val constructors = members.filter { isConstructorDeclaration(it) }
    if (constructors.isNotEmpty()) return constructors.toTypedArray()

    if (node.heritageClauses == null) return emptyArray()

    val heritageClause = node.heritageClauses?.asArray()?.find { it.token == SyntaxKind.ExtendsKeyword }
    if (heritageClause == null) return emptyArray()

    val parentReference = heritageClause.types.asArray()[0].expression

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val typeChecker = typeScriptService?.program?.getTypeChecker()
    val parentSymbol = typeChecker?.getSymbolAtLocation(parentReference)
    if (parentSymbol == null) return emptyArray()

    val parentDeclaration = parentSymbol.valueDeclaration
    if (parentDeclaration == null || !isClassDeclaration(parentDeclaration)) return emptyArray()

    val declarationMergingService = context.lookupService<DeclarationMergingService>(declarationMergingServiceKey)
    val namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    val mergedMembers = declarationMergingService
        ?.getMembers(parentDeclaration) { namespaceInfoService?.resolveNamespaceStrategy(it) }
        ?: parentDeclaration.members.asArray()

    return resolveConstructors(parentDeclaration, mergedMembers, context)
}

val convertClassDeclaration = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isClassDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    val declarationMergingService = context.lookupService<DeclarationMergingService>(declarationMergingServiceKey)
    declarationMergingService?.cover(node)

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)
    val inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
    val injectionService = context.lookupService<InjectionService>(injectionServiceKey)

    val exportModifier = node.modifiers?.asArray()?.find { it.kind === SyntaxKind.ExportKeyword }
    if (exportModifier != null) checkCoverageService?.cover(exportModifier)

    val declareModifier = node.modifiers?.asArray()?.find { it.kind === SyntaxKind.DeclareKeyword }
    if (declareModifier != null) checkCoverageService?.cover(declareModifier)

    val name = node.name?.let { render(it) } ?: "Anonymous"

    val inheritanceModifier = inheritanceModifierService?.resolveInheritanceModifier(node, context)
    val injections = injectionService?.resolveInjections(node, InjectionType.MEMBER, context, render)
    val staticInjections = injectionService?.resolveInjections(node, InjectionType.STATIC_MEMBER, context, render)
    val heritageInjections = injectionService?.resolveInjections(node, InjectionType.HERITAGE_CLAUSE, context, render)

    val namespace = typeScriptService?.findClosestNamespace(node)

    var externalModifier = "external "

    if (namespace != null && namespaceInfoService?.resolveNamespaceStrategy(namespace) == NamespaceStrategy.`object`) {
        externalModifier = ""
    }

    val typeParameters = (declarationMergingService?.getTypeParameters(node) ?: node.typeParameters?.asArray())
        ?.map { render(it) }
        ?.filter { it.isNotEmpty() }
        ?.joinToString(separator = ", ")

    val heritageClauses = (declarationMergingService?.getHeritageClauses(node) ?: node.heritageClauses?.asArray())
        ?.map { render(it) }
        ?.filter { it.isNotEmpty() }
        ?.joinToString(separator = ", ")

    val injectedHeritageClauses = heritageInjections
        ?.filter { it.isNotEmpty() }
        ?.joinToString(separator = ", ")

    val fullHeritageClauses = arrayOf(heritageClauses ?: "", injectedHeritageClauses ?: "")
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")

    val mergedMembers = declarationMergingService
        ?.getMembers(node) { namespaceInfoService?.resolveNamespaceStrategy(it) }
        ?: node.members.asArray()

    // cover private members
    mergedMembers
        .filter { member -> extractModifiers(member).any { it.kind == SyntaxKind.PrivateKeyword } }
        .forEach { member -> checkCoverageService?.cover(member) }

    val publicMembers = mergedMembers
        .filter { member -> extractModifiers(member).all { it.kind != SyntaxKind.PrivateKeyword } }

    val constructors = resolveConstructors(node, publicMembers.toTypedArray(), context)
    val otherMembers = publicMembers.filter { !isConstructorDeclaration(it) }

    val members = (constructors + otherMembers)
        .filter { member -> extractModifiers(member).all { it.kind != SyntaxKind.StaticKeyword } }
        .joinToString(separator = "\n") { render(it) }

    val staticMembers = otherMembers
        .filter { member -> extractModifiers(member).any { it.kind == SyntaxKind.StaticKeyword} }
        .joinToString(separator = "\n") { render(it) }

    val injectedMembers = (injections ?: emptyArray())
        .joinToString(separator = "\n")

    val staticInjectedMembers = (staticInjections ?: emptyArray())
        .joinToString(separator = "\n")

    var companionObject = ""

    if (staticMembers.isNotEmpty()) {
        companionObject = "\n" + """
        
companion object {
${staticMembers}${ifPresent(staticInjectedMembers) { "\n${it}" }}
}
        """.trim()
    }

    """
${ifPresent(inheritanceModifier) { "$it " }}${externalModifier}class ${name}${ifPresent(typeParameters) { "<${it}>"}}${ifPresent(fullHeritageClauses) { " : $it" }} {
${members}${ifPresent(injectedMembers) { "\n${it}" }}${companionObject}
}
    """.trim()
}
